module/zfs/spa_errlog.c

```diff
@@ -128,6 +130,103 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
 	mutex_exit(&spa->spa_errlist_lock);
 }
 
+/*
+ * If this error exists in the given tree remove it.
+ */
+static void
+remove_error_from_list(spa_t *spa, avl_tree_t *t, const zbookmark_phys_t *zb)
+{
+	spa_error_entry_t search, *found;
+	avl_index_t where;
+
+	mutex_enter(&spa->spa_errlist_lock);
+	search.se_bookmark = *zb;
+	if ((found = avl_find(t, &search, &where)) != NULL)
+		avl_remove(t, found);
+	mutex_exit(&spa->spa_errlist_lock);
+}
+
+/*
+ * If a healed bookmark matches an entry in the error log we stash it in a tree
+ * so that we can later remove the related log entries in sync context.
+ */
+static void
+spa_add_healed_error(spa_t *spa, uint64_t obj, zbookmark_phys_t *healed_zb)
+{
+	char name[NAME_MAX_LEN];
+
+	if (obj == 0)
+		return;
+
+	mutex_enter(&spa->spa_errlog_lock);
+	bookmark_to_name(healed_zb, name, sizeof (name));
+	if (zap_contains(spa->spa_meta_objset, obj, name) == 0) {
+		/*
+		 * Found an error matching healed zb, add zb to our
+		 * tree of healed errors
+		 */
+		avl_tree_t *tree = &spa->spa_errlist_healed;
+		spa_error_entry_t search;
+		spa_error_entry_t *new;
+		avl_index_t where;
+		search.se_bookmark = *healed_zb;
+		mutex_enter(&spa->spa_errlist_lock);
+		if (avl_find(tree, &search, &where) != NULL) {
+			mutex_exit(&spa->spa_errlist_lock);
+			return;
+		}
+		new = kmem_zalloc(sizeof (spa_error_entry_t), KM_SLEEP);
+		new->se_bookmark = *healed_zb;
+		avl_insert(tree, new, where);
+		mutex_exit(&spa->spa_errlist_lock);
+	}
+	mutex_exit(&spa->spa_errlog_lock);
+}
+
+/*
+ * Removes all of the recv healed errors from both on-disk error logs
+ */
+void
+spa_remove_healed_errors(spa_t *spa, dmu_tx_t *tx)
+{
+	char name[NAME_MAX_LEN];
+	spa_error_entry_t *se;
+	void *cookie = NULL;
+
+	mutex_enter(&spa->spa_errlist_lock);
+	if (avl_is_empty(&spa->spa_errlist_healed)) {
+		mutex_exit(&spa->spa_errlist_lock);
+		return;
+	}
+	while ((se = avl_destroy_nodes(&spa->spa_errlist_healed,
+	    &cookie)) != NULL) {
+		bookmark_to_name(&se->se_bookmark, name, sizeof (name));
+		kmem_free(se, sizeof (spa_error_entry_t));
+		(void) zap_remove(spa->spa_meta_objset,
+		    spa->spa_errlog_last, name, tx);
+		(void) zap_remove(spa->spa_meta_objset,
+		    spa->spa_errlog_scrub, name, tx);
+	}
+	mutex_exit(&spa->spa_errlist_lock);
+}
+
+/*
+ * Remove errors from the in-core error list and stash away bookmarks to remove
+ * them from the on-disk error log later in spa_remove_healed_errors().
+ */
+void
+spa_remove_error(spa_t *spa, zbookmark_phys_t *zb)
+{
+	char name[NAME_MAX_LEN];
+
+	bookmark_to_name(zb, name, sizeof (name));
+
+	remove_error_from_list(spa, &spa->spa_errlist_last, zb);
+	remove_error_from_list(spa, &spa->spa_errlist_scrub, zb);
+	spa_add_healed_error(spa, spa->spa_errlog_last, zb);
+	spa_add_healed_error(spa, spa->spa_errlog_scrub, zb);
```

Comment:
This logic seems overly complicated to me, why couldn't we instead add every healed bookmark to the `spa_errlist_healed` tree and then update `spa_errlog_sync` to update the on-disk errlog accordingly.  This would be very similar to how the existing `spa_errlog_last` is managed.  `spa_errlog_sync()` is already called from `spa_sync_iterate_to_convergence` so you'd want to just make sure it doesn't bail early when `avl_numnodes(&spa->spa_errlist_healed) != 0`.  Or am I missing something?